### PR TITLE
Add a notebook for partitioning

### LINKED ISSUE #73: Check if parameters of `Partioning` should be optimizable
As pointed out in [this comment](https://github.com/WUR-AI/diffWOFOST/pull/70#pullrequestreview-3656825087) we can check if the parameters of `Partitioning` can and should be optimized. 
If that is the case, we should also add a notebook to show the optimization, similar to other modules.

### PR DESCRIPTION
**Pull request by SarahAlidoost**
closes #73 

In this notebook, I implemented a different approach than the one in [pcse notebook / 11 Optimizing partitioning in a PCSE model.ipynb](https://github.com/ajwdewit/pcse_notebooks/blob/master/11%20Optimizing%20partitioning%20in%20a%20PCSE%20model.ipynb).

I will update doc after #62

### CODE REVIEW

**Review by SCiarella — CHANGES_REQUESTED** [review]
:wave: @SarahAlidoost, the notebook looks nice.

One first technical comment is that all the notebooks need to add:
```python
from diffwofost.physical_models.config import ComputeConfig
ComputeConfig.set_device('cpu')
``` 
otherwise they fail if a GPU is available because it would be the default device of the model, but not of the notebook's variables.

-----
In terms of content, I see that we are no longer optimizing a logistic function, but rather optimizing the parameters directly. To me, this seems cleaner (and it is probably faster), but maybe those sigmoid functions have a deeper meaning that we are losing? Anyway, let's wait for someone with more knowledge to comment on this issue.

---
Finally, I do not fully understand the last figure: it is just the xy coordinates of predicted vs true parameter?

**Review by ronvree — CHANGES_REQUESTED** [review]
Looks good to me! I do have some comments/questions/requests though

I noticed a `ComputeConfig` was added. I’m a bit concerned about the direction in its current form, mainly because it introduces a second “global source of truth” for dtype/device that can be easy to get out of sync with normal PyTorch conventions.
 - It keeps a global dtype that, if I understand correctly, should be the default dtype to initialize float tensors if not specified otherwise. PyTorch already has this built in (see `torch.set_default_dtype`). If these are not in sync it will cause unexpected behavior. If nothing is implemented it already behaves exactly as intended.
 - It checks whether a GPU is available and if so will use it as default, and to disable this a user has to explicitly disable this by adding `ComputeConfig.set_device('cpu')` any time a model is used and would give an error otherwise. I think it would be more intuitive to use the cpu by default, for which no extra code is required either.
 
 My impression is that PyTorch Modules generally follow the following convention
 ```
 class A(nn.Module):
 
	def __init__(self, *args, dtype=None, device=None):
		
		# pass dtype and device to any sub-modules or parameters that are initialized
		# for any initialized tensor for which dtype or device are None they are initialized to `torch.get_default_dtype` and `'cpu'`, respectively.
		...
		
	def forward(self, x):
		
		# infer dtype and device from x 
		# initialize any tensors to be consistent with dtype and device inferred from x
		... 
```
and that by following this convention no other global config would be required. 

Also in the notebook I think it would be useful to add more comments to the code to give more insight in the individual steps that are done and why it is necessary. For example, I was a bit confused about the `get_tables` output structure in `OptDiffPartitioning`. If I understand correclty it sort of interleaves the dvs and parameter values? Is this because Wofost expects the table to be parameterized this way? Some comments in these kind of steps would be super useful

**Comment by SarahAlidoost** [comment]
> Looks good to me! I do have some comments/questions/requests though
> 
> I noticed a `ComputeConfig` was added. I’m a bit concerned about the direction in its current form, mainly because it introduces a second “global source of truth” for dtype/device that can be easy to get out of sync with normal PyTorch conventions.
> 
> * It keeps a global dtype that, if I understand correctly, should be the default dtype to initialize float tensors if not specified otherwise. PyTorch already has this built in (see `torch.set_default_dtype`). If these are not in sync it will cause unexpected behavior. If nothing is implemented it already behaves exactly as intended.
> * It checks whether a GPU is available and if so will use it as default, and to disable this a user has to explicitly disable this by adding `ComputeConfig.set_device('cpu')` any time a model is used and would give an error otherwise. I think it would be more intuitive to use the cpu by default, for which no extra code is required either.
> 
> My impression is that PyTorch Modules generally follow the following convention
> 
> ```
> class A(nn.Module):
> 
>    def __init__(self, *args, dtype=None, device=None):
>    	
>    	# pass dtype and device to any sub-modules or parameters that are initialized
>    	# for any initialized tensor for which dtype or device are None they are initialized to `torch.get_default_dtype` and `'cpu'`, respectively.
>    	...
>    	
>    def forward(self, x):
>    	
>    	# infer dtype and device from x 
>    	# initialize any tensors to be consistent with dtype and device inferred from x
>    	... 
> ```
> 
> and that by following this convention no other global config would be required.
> 

@ronvree Thanks for the review and taking a close look! We think the `ComputeConfig` approach is a solid (and cleaner) alternative to using `torch.set_default_dtype/device` in diffwofost. Relying on PyTorch’s global defaults can get risky for reusable software since they’re process-wide and affect everything created afterward, including third-party code, tests, or models we don’t control. As a rule of thumb, `torch.set_default_*` is fine for notebooks or tiny scripts. For a modular code we use a config like `ComputeConfig` e.g.:
```python
dtype = ComputeConfig.get_dtype()
device = ComputeConfig.get_device()
torch.tensor(..., dtype=dtype, device=device)
```
That said, I noticed that we’re currently setting device/dtype in two places (`ComputeConfig` and `EngineTestHelper(device=..., dtype=...)`) Also, I didn’t set device/dtype correctly when creating tensors in the notebooks.  We’ll fix that in #84.

One more thing: `model.to("cuda")` won’t actually do anything here. Instead, the intended pattern is to set the device/dtype via `ComputeConfig` before running the forward pass. This approach is consistent with how other scientific simulators handle execution context.
